app/src/org/commcare/activities/CreatePinActivity.java

```diff
@@ -103,7 +103,12 @@ private void setListeners() {
             public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                 // processes the done/next keyboard action
                 if (actionId == EditorInfo.IME_ACTION_DONE || actionId == EditorInfo.IME_ACTION_NEXT) {
-                    continueButton.performClick();
+                    if (enterPinBox.getText().toString().length() != 4) {
```

Comment:
@amstone326 I think maybe I'm just missing something, but it seems confusing to me that the validation logic would be replicated here, v. unified in whatever happens when continueButton.performClick() is called. Is there a reason that pattern is hard here?

Reply:
@ctsims The reason this is implemented separately here is because for `continueButton` (i.e. the button I created on-screen, rather than the built-in keyboard action), I handled this by making that button completely disabled until the input length reaches 4 (see the implementation of `afterTextChanged` in `CreatepinActivity.getPinTextWatcher()`). This pattern is not an option for the keyboard action buttons because there is no concept of disabling them; the only method of intervention available is to override `onEditorAction` and potentially intercept what happens when it's pressed. This is also why I missed preventing this in the original implementation of this feature (this isn't a regression; it's always been this way). I could have instead placed this new code in the `onClick` methods for `continueButton` and `enterButton`, respectively, but I see no advantage to doing it that way.

Reply:
aha! That certainly clarifies things. Thanks for explaining, it was a bit confusing to tell at a glance from the code.

I think it would be useful to separate the logic such that the button press being possible and the enter button being pressed follow the same code path for validation (so that if a future constraint was added or "4" was changed it wouldn't need to be changed in both places, but Approved to pull with or without that change.